### PR TITLE
Add JinaReranker for local Jina Reranker v3 inference

### PR DESCRIPTION
**Pull request by sahel-sh**
# Pull Request Checklist

## Reference Issue

ref: N/A

## Summary

- Add `JinaReranker` class under `pointwise/` for local inference with the `jinaai/jina-reranker-v3` HuggingFace model (0.6B-param encoder-based listwise scorer).
- Smart chunk sizing: when `max_passage_words` is set, estimates how many docs fit per forward pass from `context_size` and average doc length. When unset, derives a per-doc word limit from `context_size / num_docs`.
- Hard cap at 64 docs per `model.rerank()` call (Jina v3 architecture limit).
- Jina scores are absolute — raw scores used directly across chunks (no normalization).
- Scores cast to Python `float` for JSON serialization compatibility.
- Wire into `Reranker.create_model_coordinator` factory for model paths containing `jina-reranker`.
- Add minimal `jina_template.yaml` prompt template stub.
- Add `rerank_jina.py` demo script following the `rerank_pointwise_vllm.py` pattern (argparse, BM25 retrieval on dl19, eval metrics, JSONL/TREC output).
- 25 unit tests covering chunking, raw-score preservation, float type safety, doc text extraction, input immutability, invocation history, and edge cases.

## Why

Jina Reranker v3 is a high-quality 0.6B-param multilingual reranker (61.94 nDCG@10 on BEIR) that runs locally via HuggingFace `transformers`. It uses a novel "last but not late interaction" architecture — causal self-attention between query and documents within a single context window — that doesn't fit neatly into generative listwise or single-doc pointwise categories. It produces per-document relevance scores (like pointwise) while benefiting from cross-document attention (like listwise). Placed under `pointwise/` since the output and post-processing (assign scores, sort) match that paradigm.

## Validation

```bash
.venv/bin/pre-commit run --all-files
.venv/bin/python -m unittest discover test/analysis
.venv/bin/python -m unittest discover test/evaluation
.venv/bin/python -m unittest discover test/rerank
```

All pass (84 tests total across the CI-gated suites, including 25 new Jina tests).

Tested the demo end-to-end:

| Metric | Retrieval (BM25) | Reranked (Jina v3) | Delta |
|---|---:|---:|---:|
| nDCG@10 | 0.5058 | 0.6838 | +0.1780 |
| MAP@100 | 0.2476 | 0.3379 | +0.0903 |
| Recall@20 | 0.2012 | 0.2581 | +0.0569 |
| Recall@100 | 0.4531 | 0.4531 | +0.0000 |

## Follow-ups

- Thread `--dtype` through the CLI / MCP operations layer in a separate PR.
- Consider a dedicated `scored_listwise/` subdirectory if more encoder-based multi-doc scorers are added.
- Add optional embedding extraction support (`return_embeddings=True` from Jina v3).

## Checklist Items

- [x] Have you followed the [contributing guidelines](CONTRIBUTING.md)?
- [x] Have you verified that there are no existing [Pull Requests](https://github.com/castorini/rank_llm/pulls) for the same update/change?
- [x] Have you updated any relevant documentation or added new tests where needed?

# PR Type

- [x] Feature

<div><a href="https://cursor.com/agents/bc-42c39554-d26f-4497-ac24-fda32aab3ec4"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cursor.com/assets/images/open-in-web-dark.png"><source media="(prefers-color-scheme: light)" srcset="https://cursor.com/assets/images/open-in-web-light.png">Open in Web</picture></a>&nbsp;<a href="https://cursor.com/background-agent?bcId=bc-42c39554-d26f-4497-ac24-fda32aab3ec4"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cursor.com/assets/images/open-in-cursor-dark.png"><source media="(prefers-color-scheme: light)" srcset="https://cursor.com/assets/images/open-in-cursor-light.png">Open in Cursor</picture></a>&nbsp;</div>

### CODE REVIEW

**Comment by lintool** [comment]
@codex review

File: src/rank_llm/rerank/pointwise/jina_reranker.py [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+        doc_texts_raw = [_extract_doc_text(c.doc, max_words=None) for c in candidates]
+
+        all_scores: list[float] = [0.0] * num_candidates
+
+        chunk_start = 0
+        while chunk_start < num_candidates:
+            remaining = num_candidates - chunk_start
+            target_docs = min(remaining, self._window_size)
+            effective_max_words = self._fit_max_passage_words(self._window_size)
+            chunk_size = target_docs
```
[the remark this answers is not in the document]
Comment (sahel-sh) [reply src/rank_llm/rerank/pointwise/jina_reranker.py:281]: context_size is in orders of 10-100k, per doc over head < 50. This is not an issue
